Woop installer: redirect from checkout to transfer step

retrofox opened the pull request:
#### Changes proposed in this Pull Request

This PR changes the checkout `redirect_to` URL when the site doesn't have the proper plan to `transfer`.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test with a Simple site without a proper plan (Free, for instance)
* Confirm clicking on the checkout cancel button (cross at top left), redirects to `transfer` and then redirects to `confirm` step on the fly.
* Confirm once the checkout is done is redirects straight to transfer step, avoiding the need to click on the `Sounds Good` button

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

retrofox committed:
woop: from checkout redirect to transfer

DustyReagan commented:
After checkout I still had to confirm the domain name change and click "Sounds good". I just checked-out the PR and tested. Anything else I should be sand-boxing maybe?

lsl committed:
Add siteConfirmed dependency and transfer to confirm redirect

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -185,7 +161,33 @@ export default function Confirm( props: WooCommerceInstallProps ): ReactElement
+							<StyledNextButton
+								disabled={ hasBlockers || ! isDataReady }
+								onClick={ () => {
+									dispatch(
+										submitSignupStep( { stepName: 'confirm' }, { siteConfirmed: siteId } )
+									);
+									if ( siteUpgrading.required ) {
+										return ( window.location.href = siteUpgrading.checkoutUrl );
```
This seems to work but please test with a few attempts, can this redirect break the dispatch or is that all in sync?

One of my tests got a checkout errorNotice complaining about the plan selection being undefined. Guessing maybe old site not compatible with the "business-bundle" plan selection?

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -138,33 +141,6 @@ export default function Confirm( props: WooCommerceInstallProps ): ReactElement
-			</>
-		);
-	}
-
```
Moved due to TS complaining about `dispatch` typing

retrofox replied:
🤔

lsl commented:
Added that `siteConfirmed` stuff from way back in https://github.com/Automattic/wp-calypso/pull/58536 and made the transfer step redirect to confirm if something shows up without it. This seems to have the desired effect I think. (@DustyReagan?)

Non business:
- confirm => checkout => cancel back to transfer, redirects to confirm => confirm again => complete checkout => transfer starts 
- confirm => complete checkout => transfer starts 

Non atomic:
- confirm => transfer starts

Atomic:
- confirm => install starts

retrofox commented:
The Cancel button doesn't work as expected:


https://user-images.githubusercontent.com/77539/146195763-3e899229-61c9-4700-b664-c7dd994165cd.mov

Rolilink commented:
> The Cancel button doesn't work as expected.

This seems to be happening because when the user hits the cancel button the user is directed to the transfer page without a business plan. I am adding a quick verification on the `Transfer` component and redirecting when an upgrade is required.

Rolilink committed:
woop: adds plan verification to the transfer step

Rolilink committed:
woop: adding the required part

lsl committed:
Rework siteConfirmed and things

lsl committed:
Add cancel_to to checkout

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -151,8 +154,10 @@ export default function Confirm( props: WooCommerceInstallProps ): ReactElement
+								dispatch( submitSignupStep( { stepName: 'confirm' }, { siteConfirmed: siteId } ) );
 								if ( siteUpgrading.required ) {
-									return ( window.location.href = siteUpgrading.checkoutUrl );
+									page( siteUpgrading.checkoutUrl );
```
`page` uses a `setTimeout( () =>{}, 0);` which seems to solve the race condition for dispatch not being stored before the redirect happens.

pablinos replied:
Ah, I've added that myself in the past when using `window.location.href`. Good to know that `page` handles it for us.

lsl commented on client/signup/steps/woocommerce-install/transfer/index.tsx:
```diff
@@ -14,6 +14,16 @@ export default function Transfer( props: WooCommerceInstallProps ): ReactElement
+		signupDependencies: { siteConfirmed },
+	} = props;
+
+	if ( siteConfirmed !== siteId ) {
```
Checking this isn't *that* important in the end (dispatching `submitSignupStep` was the real fix) but because state gets stored I want to rule out any potential resume problems.

lsl commented on client/my-sites/checkout/composite-checkout/composite-checkout.tsx:
```diff
@@ -700,14 +700,23 @@ export default function CompositeCheckout( {
 				const redirectPath = searchParams.get( 'redirect_to' ) ?? '';
 				// Only allow redirecting to relative paths.
-				if ( redirectPath.startsWith( '/' ) ) {
+				if ( redirectPath.match( /^\/[^/]?/ ) ) {
```
@sirbrillig / @michaeldcain I don't think we're using the composite checkout here but noticed the matching code in masterbar/checkout.tsx. This is to prevent `//example.com` getting treated as safe to redirect.

Can you confirm this code only runs for the "cancel" path? `redirect_to` looks like its handled somewhere else for the "checkout success" path.

Also, how can I test the composite checkout behaves the same / as expected after this change? Is it in use yet? Readme had a note about this still being under devel.

pablinos replied:
In the interest of getting this merged, could we remove the `cancel_to` change for now? When they cancel, the site won't have the correct plan and should be redirected back to the confirm step.

michaeldcain replied:
> Can you confirm this code only runs for the "cancel" path? redirect_to looks like its handled somewhere else for the "checkout success" path.

That's correct. The post-Checkout handling of `redirect_to` happens [in `getThankYouPageURL()`](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L94). I think this code path only runs for the empty cart view that happens when there's an error adding an item to the shopping cart. You can test it by [setting this to `true`](https://github.com/Automattic/wp-calypso/blob/fix%2Fpurchase-meta-alignment/client/my-sites/checkout/composite-checkout/composite-checkout.tsx#L730) and visiting Checkout.

I also tested the similar change to `masterbar/checkout.tsx` and it works as expected, so this LGTM. Thanks for digging deeper to keep them in sync.

> Is it in use yet? Readme had a note about this still being under devel.

Composite Checkout is the only version of Checkout now. Thus, if you buy a Jetpack or WPcom product, you're testing it :). I'll update that README.

pablinos replied:
Excellent, thanks @michaeldcain. It sounds like this could be ready to merge with a bit more testing then.

Just noting that one of the links in your comment was broken for me, but [it seemed to be pointing here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/composite-checkout.tsx#L730), which makes sense

sirbrillig commented on client/layout/masterbar/checkout.tsx:
```diff
@@ -55,14 +55,23 @@ const CheckoutMasterbar: FunctionComponent< Props > = ( {
 				const redirectPath = searchParams.get( 'redirect_to' ) ?? '';
 				// Only allow redirecting to relative paths.
-				if ( redirectPath.startsWith( '/' ) ) {
+				if ( redirectPath.match( /^\/[^/]?/ ) ) {
```
This code is being moved to a stand-alone function in https://github.com/Automattic/wp-calypso/pull/59226 so you'll probably need a rebase after that merges, but the good news is that it will affect all three places that deal with closing checkout!

pablinos replied:
Thanks @sirbrillig. I've just rebased this after #59226 was merged. Everything seems ok still. Would you mind giving it a quick look?

sirbrillig reviewed (commented):
I haven't tested this, nor can I speak to the woocommerce-install parts, but the checkout parts look fine to me!

DustyReagan reviewed (approved):
Looks great! 👍

https://user-images.githubusercontent.com/140841/146461698-068db5a6-a338-4ceb-932d-ceb153a0ade3.mp4

lsl committed:
Fix regex

I was trying to allow "/" routes by making the second char requirement
optional. Needed negative lookahead instead.

lsl commented:
Had to fix the regex, retested the installer flows and everything works as expected.

https://user-images.githubusercontent.com/811776/146475873-94a9a066-4aa8-423e-9042-8852224e3ff1.mp4

Also tested standard checkout flows from /start and /plans and that all still work as expected.

jsnajdr commented on client/me/help/help-contact/index.jsx:
```diff
@@ -103,7 +103,7 @@ class HelpContact extends Component {
 		const searchParams = new URLSearchParams( window.location.search );
 		const redirectPath = searchParams.get( 'redirect_to' ) ?? '';
-		if ( redirectPath.match( /^\/[^/]?/ ) ) {
+		if ( redirectPath.match( /^\/(?!\/)/ ) ) {
```
This regexp match is validating whether the `redirect_to` URL is valid, e.g., doesn't contain some malicious `javascript:` snippet or redirect to some completely different site? And does that by checking if the path starts with `/` but doesn't start with `//`?

There's a better, more general-purpose way to do this check:
```js
function check( url ) {
  try {
    return new URL( url, window.location.href ).origin === window.location.origin;
  } catch {
    return false;
  }
}
```
It checks whether the link is a valid URL that is same-origin, i.e., inside the same site, and accepts also relative links like `help` or `./help` or `../help`, ones which would work well with `window.location.href = link`.

This could be extracted a to a common `lib` helper, as there are plenty of usage opportunites in Calypso. And making redirect validation as easy as possible looks like a great idea.
